site/js/modules/webui/input.js

```diff
@@ -0,0 +1,187 @@
+            } = {}) {
+        // The name of the field which will be used for targeting the field.
+        this.name = name;
+
+        // The display name that will be shown to the user.
+        this.displayName = displayName;
+
+        // The HTML input type.
+        // Supports checkbox, email, number, password, and text types.
+        this.type = type;
+
+        // The underlying type of the field.
+        // Non-standard types are parsed to JSON.
+        // If a non-standard type cannot be parsed to JSON,
+        // a validation error is raised.
+        this.underlyingType = underlyingType;
+
+        // Flags the field requires user input.
+        this.required = required;
+
+        // The placeholder text for the input.
+        this.placeholder = placeholder;
+
+        // Optional classes that are attached to the input.
+        this.inputClass = inputClass;
+
+        // Any additional attributes to the input field.
+        // If the field is required, the required attribute will be added automatically.
+        this.attributes = attributes;
+
+        // Determines the position of the HTML label with respect to the input.
+        this.labelBefore = labelBefore;
+
+        // A custom function for extracting the value from an input.
+        // If a default value extraction cannot be inferred, the value is parsed using JSON.
+        this.extractInputFunc = extractInputFunc;
+    }
+
+    validate() {
+        if ((this.name == undefined) || (this.name == '')) {
+            console.error(`Input field cannot have an empty name: ${JSON.stringify(this)}.`);
+        }
+
+        if ((this.displayName == undefined) || (this.displayName == '')) {
+            console.error(`Input field cannot have an empty display name: ${JSON.stringify(this)}.`);
+        }
+    }
+
+    toHTML() {
+        this.validate();
+
+        let attributes = this.attributes;
+        let displayName = this.displayName;
+        if ((this.required) && (this.placeholder === "")) {
+            attributes += ' required';
+            displayName += ` <span class="required-color">*</span>`;
+        }
+
+        let inputFieldHTML = [
+            `<label for="${this.name}">${displayName}</label>`,
+            `<input type="${this.type}" id="${this.name}" name="${this.name}" placeholder="${this.placeholder}" ${attributes}/>`,
+        ];
+
+        if (!this.labelBefore) {
+            inputFieldHTML.reverse();
+        }
+
+        return `
+            <div class="input-field ${this.inputClass}">
+                ${inputFieldHTML.join("\n")}
+            </div>
+        `;
+    }
+
+    getResult(container) {
+        let input = container.querySelector(`fieldset [name=${this.name}]`);
+        input.classList.add("touched");
+
+        return new Result(input, this.underlyingType, this.extractInputFunc);
+    }
+}
+
+// The result of getting an Input.Field.
+class Result {
```

Comment:
This name is confusing to me since the value has not been extracted yet.
Maybe you can do something like rename the other class something like "FieldType" and this one "FieldInstance"?
Whatever you decide, "Result" is too confusing.